l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,742 @@
+
+        if ((_chainId == ERA_CHAIN_ID) && ((_l2BatchNumber < eraIsEthWithdrawalFinalizedStorageSwitchBatchNumber))) {
+            // in this case we have to check we don't double withdraw ether
+            // we are not fully finalized if eth has not been withdrawn
+            // note the WETH bridge has not yet been deployed, so it cannot be the case that we withdrew Eth but not WETH.
+            bool alreadyFinalized = IGetters(ERA_DIAMOND_PROXY).isEthWithdrawalFinalized(
+                _l2BatchNumber,
+                _l2MessageIndex
+            );
+            require(!alreadyFinalized, "Withdrawal is already finalized");
+        }
+
+        bool wrapToWeth;
+
+        {
+            MessageParams memory messageParams = MessageParams({
+                l2BatchNumber: _l2BatchNumber,
+                l2MessageIndex: _l2MessageIndex,
+                l2TxNumberInBatch: _l2TxNumberInBatch
+            });
+
+            (l1Receiver, l1Token, amount, wrapToWeth) = _checkWithdrawal(
+                _chainId,
+                messageParams,
+                _message,
+                _merkleProof
+            );
+        }
+
+        if (!hyperbridgingEnabled[_chainId]) {
+            // Check that the chain has sufficient balance
+            require(chainBalance[_chainId][l1Token] >= amount, "EB n funds 2"); // not enought funds
+            chainBalance[_chainId][l1Token] -= amount;
+        }
+
+        if (wrapToWeth) {
+            // Wrap ETH to WETH tokens (smart contract address receives the equivalent amount of WETH)
+            IWETH9(l1WethAddress).deposit{value: amount}();
+            // Transfer WETH tokens from the smart contract address to the withdrawal receiver
+            IERC20(l1WethAddress).safeTransfer(l1Receiver, amount);
+
+            emit EthWithdrawalFinalized(_chainId, l1Receiver, amount);
+        } else if ((l1Token == ETH_TOKEN_ADDRESS) || (l1Token == l1WethAddress)) {
+            bool callSuccess;
+            // Low-level assembly call, to avoid any memory copying (save gas)
+            assembly {
+                callSuccess := call(gas(), l1Receiver, amount, 0, 0, 0, 0)
+            }
+            require(callSuccess, "L1WB: withdraw failed");
+            emit EthWithdrawalFinalized(_chainId, l1Receiver, amount);
+        } else {
+            // Withdraw funds
+            IERC20(l1Token).safeTransfer(l1Receiver, amount);
+
+            emit WithdrawalFinalizedSharedBridge(_chainId, l1Receiver, l1Token, amount);
+        }
+    }
+
+    /// @dev check that the withdrawal is valid
+    function _checkWithdrawal(
+        uint256 _chainId,
+        MessageParams memory _messageParams,
+        bytes calldata _message,
+        bytes32[] calldata _merkleProof
+    ) internal view returns (address l1Receiver, address l1Token, uint256 amount, bool wrapToWeth) {
+        (l1Receiver, l1Token, amount, wrapToWeth) = _parseL2WithdrawalMessage(_chainId, _message);
```

Comment:
this code will be run when trying to finalize withdrawals from the legacy Era also, so not sure if it will work

Reply:
We made the messaging backwards compatible, so it should work